### PR TITLE
Add `IVTable` interfaces to COM structs

### LINKED ISSUE #831: Add IVTable interface to allow programmatically populating CCW
**Is your feature request related to a problem? Please describe.**
For a given COM wrapping struct (say `IStream`) it would be good to have the ability to allocate and populate a vtable utilizing the new `PopulateVTable` method. In WinForms we have to accomplish this by making partials for every struct to apply this interface.

**Describe the solution you'd like**
By applying an `IVTable` interface that leverages static interface methods, structs can be utilized via constrained generic helper methods without knowing which specific struct is being used. As a concrete example- to populate a `ComInterfaceEntry` table for `ComWrappers` you could write the following:

``` C#
// Called like so: var table = ComInterfaceTable.Create<IStream>();
public static ComInterfaceTable Create<TComInterface>()
    where TComInterface : unmanaged, IComIID, IVTable
{
    Span<ComInterfaceEntry> entries = AllocateEntries<TComInterface>(1);
    entries[0] = new()
    {
        Vtable = (nint)TComInterface.GetVTable(),
        IID = *IID.Get<TComInterface>()
    };

    return new()
    {
        Entries = (ComInterfaceEntry*)Unsafe.AsPointer(ref entries[0]),
        Count = entries.Length
    };
}
```

## Proposed implementation ##

This is based on what WinForms currently has implemented.

``` C#
/// <summary>
///  Non generic interface that allows constraining against a COM wrapper type directly. COM structs should
///  implement <see cref="IVTable{TComInterface, TVTable}"/>.
/// </summary>
internal unsafe interface IVTable
{
    static abstract IUnknown.Vtbl* GetVTable();
}

internal unsafe interface IVTable<TComInterface, TVTable> : IVTable
    where TVTable : unmanaged
    where TComInterface : unmanaged, IComIID, IVTable<TComInterface, TVTable>
{
    private static sealed TVTable* VTable { get; set; }

    private protected static abstract void PopulateVTable(TVTable* vtable);

    static IUnknown.Vtbl* IVTable.GetVTable()
    {
        if (VTable is null)
        {
            TVTable* vtable = (TVTable*)RuntimeHelpers.AllocateTypeAssociatedMemory(typeof(TVTable), sizeof(TVTable));
            ComHelpers.PopulateIUnknown<TComInterface>((IUnknown.Vtbl*)vtable);
            VTable = vtable;
            TComInterface.PopulateVTable(VTable);
        }

        return (IUnknown.Vtbl*)VTable;
    }
}
```

`ComHelpers` would be extended (by CsWin32) as follows:

``` C#
static unsafe partial class ComHelpers
{
    public static void PopulateIUnknown<TComInterface>(IUnknown.Vtbl* vtable)
        where TComInterface : unmanaged, IComIID
        => PopulateIUnknownImpl<TComInterface>(vtable);

    static partial void PopulateIUnknownImpl<TComInterface>(IUnknown.Vtbl* vtable) where TComInterface : unmanaged, IComIID;
}
```

User code, such as WinForms, would extend the partial class like this to attach `IUnknown` callbacks. (These callbacks are protected in `ComWrappers`.)

``` C#
static unsafe partial class ComHelpers
{
    static partial void PopulateIUnknownImpl<TComInterface>(IUnknown.Vtbl* vtable) where TComInterface : unmanaged, IComIID
    {
        // Custom behavior for specific COM CCWs can be done by checking typeof(TComInterface) against a specific
        // interface, such as typeof(IAccessible).
        CustomComWrappers.PopulateIUnknown(vtable);
    }
}
```

**Additional context**
This is a building block that we're able to use to apply another similar pattern to implementations of one of the CsWin32 generated COM struct interfaces to easily enlist it for exposure to COM through COM wrappers. To expose an `IStream` all we need to do is implement `IStream.Interface` and add a single additional interface:

``` C#
public class MyStream : IStream.Interface, IManagedWrapper<IStream, ISequentialStream>
{
    // IStream.Interface implementation
}
```

### PR DESCRIPTION
**Pull request by AArnott**
Closes #831

### CODE REVIEW

File: test/GenerationSandbox.Unmarshalled.Tests/ComHelpers.cs [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    private static uint AddRef(IUnknown* punk)
+    {
+        throw new NotImplementedException();
+    }
+
+    [UnmanagedCallersOnly(CallConvs = new[] { typeof(CallConvStdcall) })]
+    private static uint Release(IUnknown* punk)
+    {
+        throw new NotImplementedException();
+    }
```
Comment (AArnott) [inline test/GenerationSandbox.Unmarshalled.Tests/ComHelpers.cs:42]: @jeremykuhne For purposes of a mini-sample, I'd like to actually implement this properly (with ComWrappers) if possible. But I couldn't figure out how. Can you help?

File: src/Microsoft.Windows.CsWin32/templates/ComHelpers.cs [hunk 2]
```diff
@@ -10,4 +10,19 @@
...
+
+#if canUseInterfaceStaticMembers
+	internal static void PopulateIUnknown<TComInterface>(System.Com.IUnknown.Vtbl* vtable)
+		where TComInterface : unmanaged
+	{
+		PopulateIUnknownImpl<TComInterface>(vtable);
+		if (vtable->QueryInterface_1 is null)
+		{
+			throw new NotImplementedException("v-tables cannot be accessed unless the Windows.Win32.ComHelpers.PopulateIUnknownImpl partial method is implemented.");
+		}
```
Comment (AArnott) [inline src/Microsoft.Windows.CsWin32/templates/ComHelpers.cs:22]: This is the extent of our runtime check to ensure that if the user wants the vtable that they have implemented the partial method.

File: test/GenerationSandbox.Unmarshalled.Tests/ComHelpers.cs (hunk 1 above)
Comment (JeremyKuhne) [reply test/GenerationSandbox.Unmarshalled.Tests/ComHelpers.cs:42]: Sure. To just get it to populate the IUnknown entries you can do this:

``` C#
public class TestComWrappers : ComWrappers
{
    private static readonly IUnknown.Vtbl* s_wrappersUnknown = GetComWrappersUnknown();

    private static IUnknown.Vtbl* GetComWrappersUnknown()
    {
        IUnknown.Vtbl* vtable = (IUnknown.Vtbl*)RuntimeHelpers.AllocateTypeAssociatedMemory(
            typeof(CustomComWrappers),
            sizeof(IUnknown.Vtbl));

        GetIUnknownImpl(out nint fpQueryInterface, out nint fpAddRef, out nint fpRelease);
        vtable->QueryInterface_1 = (delegate* unmanaged[Stdcall]<IUnknown*, Guid*, void**, HRESULT>)fpQueryInterface;
        vtable->AddRef_2 = (delegate* unmanaged[Stdcall]<IUnknown*, uint>)fpAddRef;
        vtable->Release_3 = (delegate* unmanaged[Stdcall]<IUnknown*, uint>)fpRelease;

        return vtable;
    }

    // Call this to populate the Vtbl
    internal static void PopulateIUnknown(IUnknown.Vtbl* vtable)
    {
        vtable->QueryInterface_1 = s_wrappersUnknown->QueryInterface_1;
        vtable->AddRef_2 = s_wrappersUnknown->AddRef_2;
        vtable->Release_3 = s_wrappersUnknown->Release_3;
    }

    // Abstracts that need implementation
    protected override unsafe ComInterfaceEntry* ComputeVtables(object obj, CreateComInterfaceFlags flags, out int count)
    {
        count = 0;
        return null;
    }

    protected override object? CreateObject(nint externalComObject, CreateObjectFlags flags) => null;
    protected override void ReleaseObjects(IEnumerable objects) => throw new NotImplementedException();
}
```

Creating a full CCW requires implementing `ComputeVtables`.

File: src/Microsoft.Windows.CsWin32/templates/ComHelpers.cs (hunk 2 above)
Comment (JeremyKuhne) [reply src/Microsoft.Windows.CsWin32/templates/ComHelpers.cs:22]: That should be good, imo.

**Pull request merged**